### PR TITLE
Distinguished bottom from stuck evaluation

### PR DESCRIPTION
**Pull request by traiansf**
Ready for review.

###### Reviewer checklist

- [x] Test coverage: `stack test --coverage`
- [x] Public API documentation: `stack haddock`
- [x] Style conformance: `stylish-haskell`

---

### CODE REVIEW

File: src/main/haskell/kore/src/Kore/Variables/Fresh/IntCounter.hs [hunk 1]
```diff
@@ -38,3 +39,20 @@ instance Monad IntCounter where
 instance MonadState Int IntCounter where
     get = IntCounter get
     put = IntCounter . put
+
+{-|@fisrtNotFalse@ takes a predicate and a list of 'MonadState' actions, and
+locates the first action whose result satisfies the predicate, resetting the
+state before executing each action.
+-}
+firstNotFalse
```
Comment (virgil-serbanuta) [inline src/main/haskell/kore/src/Kore/Variables/Fresh/IntCounter.hs]: `Data.List` seems to call this `find`. Why not `findState`?

File: src/main/haskell/kore/src/Kore/Step/BaseStep.hs [hunk 2]
```diff
@@ -237,30 +244,38 @@ stepWithAxiom
         , substitution
         , _  -- TODO: Use this proof
         ) <-
-            normalizeUnificationError existingVars
+            normalizeUnificationError
+                (makeFalsePredicate, [], EmptyUnificationProof)
```
Comment (virgil-serbanuta) [inline src/main/haskell/kore/src/Kore/Step/BaseStep.hs]: Test this?

File: src/main/haskell/kore/src/Kore/Unification/Error.hs [hunk 3]
```diff
@@ -54,6 +59,10 @@ mapSubstitutionErrorVariables
     :: (variableFrom level -> variableTo level)
     -> SubstitutionError level variableFrom
     -> SubstitutionError level variableTo
-mapSubstitutionErrorVariables mapper (CircularVariableDependency variables) =
-    CircularVariableDependency (map mapper variables)
+mapSubstitutionErrorVariables mapper
+    (CtorCircularVariableDependency variables) =
+        CtorCircularVariableDependency (map mapper variables)
```
Comment (virgil-serbanuta) [inline src/main/haskell/kore/src/Kore/Unification/Error.hs]: This line is not tested.

File: src/main/haskell/kore/src/Kore/Predicate/Predicate.hs [hunk 4]
```diff
@@ -120,6 +121,12 @@ pattern PredicateTrue :: Predicate level var
 pattern PredicateFalse <- GenericPredicate(Bottom_ _)
 pattern PredicateTrue <- GenericPredicate(Top_ _)
 
+{-|'isFalsePredicate' checks whether a predicate matches 'PredicateFalse'
```
Comment (virgil-serbanuta) [inline src/main/haskell/kore/src/Kore/Predicate/Predicate.hs]: Dot at the end.

File: src/main/haskell/kore/src/Kore/Unification/Error.hs [hunk 5]
```diff
@@ -33,8 +33,11 @@ data UnificationError level
 {--| 'SubstitutionError' specifies the various error cases related to
 substitutions.
 --}
-newtype SubstitutionError level variable =
-    CircularVariableDependency [variable level]
+data SubstitutionError level variable
+    = CtorCircularVariableDependency [variable level]
+    -- ^the circularity path passes only through constructors: unsolvable
```
Comment (virgil-serbanuta) [inline src/main/haskell/kore/src/Kore/Unification/Error.hs]: Dot at the end? Here and below.

File: src/main/haskell/kore/src/Kore/Step/BaseStep.hs (hunk 2 above)
Comment (traiansf) [reply src/main/haskell/kore/src/Kore/Step/BaseStep.hs]: It will be tested once this is solved: https://trello.com/c/dTi0lICM

**Pull request merged**